core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,447 @@
+  private Throwable mPacketWriteException = null;
+  @GuardedBy("mLock")
+  private boolean mDone = false;
+  @GuardedBy("mLock")
+  private boolean mEOFSent = false;
+  /** This condition is met if mPacketWriteException != null or mDone = true. */
+  private Condition mDoneOrFailed = mLock.newCondition();
+  /** This condition is met if mPacketWriteException != null or the buffer is not full. */
+  private Condition mBufferNotFullOrFailed = mLock.newCondition();
+  /** This condition is met if there is nothing in the netty buffer. */
+  private Condition mBufferEmptyOrFailed = mLock.newCondition();
+
+  /**
+   * Creates an instance of {@link NettyPacketWriter}.
+   *
+   * @param context the file system context
+   * @param address the data server network address
+   * @param id the block ID or UFS file ID
+   * @param length the length of the block or file to write, set to Long.MAX_VALUE if unknown
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException it fails to acquire a netty channel
+   */
+  public NettyPacketWriter(FileSystemContext context, final InetSocketAddress address, long id,
+      long length, long sessionId, Protocol.RequestType type) throws IOException {
+    mContext = context;
+    mAddress = address;
+    mSessionId = sessionId;
+    mId = id;
+    mRequestType = type;
+    mLength = length;
+    mChannel = context.acquireNettyChannel(address);
+
+    ChannelPipeline pipeline = mChannel.pipeline();
+    if (!(pipeline.last() instanceof RPCMessageDecoder)) {
+      throw new RuntimeException(String
+          .format("Channel pipeline has unexpected handlers %s (expects RPCMessageDecoder).",
+              pipeline.last().getClass().getCanonicalName()));
+    }
+    mChannel.pipeline().addLast(new PacketWriteHandler());
+  }
+
+  @Override
+  public long pos() {
+    mLock.lock();
+    try {
+      return mPosToQueue;
+    } finally {
+      mLock.unlock();
+    }
+  }
+
+  @Override
+  public void writePacket(final ByteBuf buf) throws IOException {
+    final long len;
+    final long offset;
+    mLock.lock();
+    try {
+      Preconditions.checkState(!mClosed && !mEOFSent);
+      Preconditions.checkArgument(buf.readableBytes() <= PACKET_SIZE);
+      while (true) {
+        if (mPacketWriteException != null) {
+          throw new IOException(mPacketWriteException);
+        }
+        if (!tooManyPacketsInFlight()) {
+          offset = mPosToQueue;
+          mPosToQueue += buf.readableBytes();
+          len = buf.readableBytes();
+          break;
+        }
+        try {
+          if (!mBufferNotFullOrFailed.await(WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+            throw new IOException(
+                String.format("Timeout to write packet to %d @ %s.", mId, mAddress));
+          }
+        } catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    } catch (Throwable e) {
+      buf.release();
```

Comment:
Yes, we do need to release as we increment the reference count when the buf is passed to the PacketWriter.

The general rule is that the consumer always need to release the buf. 

I increment the reference count when passing it to PacketWriter because there might be more than 1 PacketWriters for the feature you developed.